compiler/ballerina-lang/src/main/java/io/ballerina/projects/PackageCompilation.java

```diff
@@ -210,4 +221,73 @@ private static void addCompilerPluginDiagnostics(PackageCompilation compilation,
         //  available to the build command
         compilation.pluginDiagnostics.addAll(reportedDiagnostics);
     }
+
+    /**
+     * Report if non exported module is imported in the current package.
+     *
+     * @param compilation package compilation
+     */
+    private static void reportNonExportedModules(PackageCompilation compilation) {
+        ResolvedPackageDependency resolvedPackageDependency = new ResolvedPackageDependency(
+                compilation.rootPackageContext.project().currentPackage(), PackageDependencyScope.DEFAULT);
+        Collection<ResolvedPackageDependency> directDependencies = compilation.packageResolution.dependencyGraph()
+                .getDirectDependencies(resolvedPackageDependency);
+
+        List<Diagnostic> allDiagnostics = new ArrayList<>(compilation.diagnosticResult.diagnostics());
+        List<Diagnostic> nonExportedModulesDiagnostics = new ArrayList<>();
+
+        for (ModuleId moduleId: compilation.rootPackageContext.project().currentPackage().moduleIds()) {
+            Module module = compilation.rootPackageContext.project().currentPackage().module(moduleId);
+
+            if (DOT.equals(module.moduleName().toString())) {
+                continue;
+            }
+
+            for (DocumentId documentId : module.documentIds()) {
```

Comment:
This is an expensive loop. We are iterating through all the documents first and each import declaration in a document, then we iterate through all the dependencies of the current package. 

I think we need to find a better way to solve this.